File: src/ios/MobihelpPlugin.m
//
//  MobihelpPlugin.m
//  FreshdeskSDK
//
//  Copyright (c) 2014 Freshdesk. All rights reserved.
//

#import <Cordova/CDV.h>

#import "MobihelpPlugin.h"
#import "Mobihelp.h"




//Implementation for the mobihelp iOS Plugin
@implementation MobihelpPlugin:CDVPlugin




//Utility methods
-(void) callbackToJavascriptWithResult:(CDVPluginResult*)result ForCommand:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

-(void) callbackToJavascriptWithoutResultForCommand :(CDVInvokedUrlCommand*)command {
    CDVPluginResult* emptyResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self callbackToJavascriptWithResult:emptyResult ForCommand:command];
}
-(void) callbackToJavascriptWithException :(NSException*)e ForCommand:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[e name]];
    [self callbackToJavascriptWithResult:result ForCommand:command];
}

-(void) logToDevice:(CDVInvokedUrlCommand*) command {
    NSArray* arguments = [command arguments];
    id toLog = [arguments objectAtIndex:0];
    NSLog(@"%@",toLog);
}
-(NSInteger)mapWithEnum :(NSString*)str {
    if ([str isEqualToString:@"FEEDBACK_TYPE_ANONYMOUS"]) {
        return FEEDBACK_TYPE_ANONYMOUS;
    } else if ([str isEqualToString:@"FEEDBACK_TYPE_NAME_AND_EMAIL_REQUIRED"]) {
        return FEEDBACK_TYPE_NAME_AND_EMAIL_REQUIRED;
    } else if ([str isEqualToString:@"FEEDBACK_TYPE_NAME_REQUIRED_AND_EMAIL_OPTIONAL"]) {
        return FEEDBACK_TYPE_NAME_REQUIRED_AND_EMAIL_OPTIONAL;
    } else {
        return -1;
    }
}




//Mobihelp methods
-(void)init :(CDVInvokedUrlCommand*)command {
    NSArray* arguments = [command arguments];
    NSDictionary* initParams = [arguments objectAtIndex:0];
    
    
    //Parse input
    NSString* domain = [initParams objectForKey:@"domain"];
    NSString* appKey = [initParams objectForKey:@"appKey"];
    NSString* appSecret = [initParams objectForKey:@"appSecret"];
    
    BOOL autoReplyEnabled = [[initParams objectForKey:@"autoReplyEnabled"] boolValue];
    BOOL enhancedPrivacyModeEnabled = [[initParams objectForKey:@"enhancedPrivacyModeEnabled"] boolValue];
    BOOL prefetchSolutions = [[initParams objectForKey:@"prefetchSolutions"] boolValue];
    NSString* feedbackType = [initParams objectForKey:@"feedbackType"];
    NSString* iosAppStoreId = [initParams objectForKey:@"iosAppStoreId"];
    NSString* iosThemeName = [initParams objectForKey:@"iosThemeName"];
    int launchCountForReviewPrompt = (int)[initParams objectForKey:@"launchCountForReviewPrompt"];

    
    //initialize
    MobihelpConfig *config = [[MobihelpConfig alloc] initWithDomain:domain withAppKey:appKey andAppSecret:appSecret];
    [[Mobihelp sharedInstance] initWithConfig:config];
    
    
    //config
    if(autoReplyEnabled) [config setEnableAutoReply:(BOOL)autoReplyEnabled];
    if(enhancedPrivacyModeEnabled) [config setEnableEnhancedPrivacy:(BOOL)enhancedPrivacyModeEnabled];
    if(prefetchSolutions) [config setPrefetchSolutions:(BOOL)prefetchSolutions];
    [config setFeedbackType: [self mapWithEnum:feedbackType]];
    [config setAppStoreId:iosAppStoreId];
    [config setThemeName:iosThemeName];
    [config setLaunchCountForAppReviewPrompt:launchCountForReviewPrompt];
}
-(void) addCustomData:(CDVInvokedUrlCommand*) command {
    NSArray* arguments = [command arguments];
    
    
    @try {
        
        
        if([arguments count]<1)
            [NSException raise: @"KeyNotFound" format:@"Key not provided for Custom Data"];
        if([arguments count]<2)
            [NSException raise: @"ValueNotFound" format:@"Value not provided for Custom Data"];
        
        NSString* key = [arguments objectAtIndex:0];
        NSString* value = [arguments objectAtIndex:1];
        bool sensitivity;
        if([arguments count]<3)
            sensitivity = false;
        else
            sensitivity = [[arguments objectAtIndex:2] boolValue];
        
        [[Mobihelp sharedInstance] addCustomDataForKey:key withValue:value andSensitivity:sensitivity];
        [self callbackToJavascriptWithoutResultForCommand:command];
    }
    @catch(NSException* e) {
        [self callbackToJavascriptWithException:e ForCommand:command];
    }
}
-(void) clearBreadCrumbs :(CDVInvokedUrlCommand*)command {
    [[Mobihelp sharedInstance] clearBreadcrumbs];
    [self callbackToJavascriptWithoutResultForCommand:command];
}
-(void) clearCustomData :(CDVInvokedUrlCommand*)command {
    [[Mobihelp sharedInstance] clearCustomData];
    [self callbackToJavascriptWithoutResultForCommand:command];
}
-(void) clearUserData :(CDVInvokedUrlCommand*)command {
    [[Mobihelp sharedInstance] clearUserData];
    [self callbackToJavascriptWithoutResultForCommand:command];
}
-(void) getUnreadCount :(CDVInvokedUrlCommand*)command {
    NSInteger unreadCount = [[Mobihelp sharedInstance] unreadCount];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsInt:(int)unreadCount];
    [self callbackToJavascriptWithResult:result ForCommand:command];
}
-(void) getUnreadCountAsync :(CDVInvokedUrlCommand*)command {
    [[Mobihelp sharedInstance] unreadCountWithCompletion:^(NSInteger unreadCount) {
        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsInt:(int)unreadCount];
        [self callbackToJavascriptWithResult:result ForCommand:command];
    }];
}
-(void) leaveBreadCrumb :(CDVInvokedUrlCommand*)command {
    @try {
        NSString* crumbDetails = [command argumentAtIndex:0];
        [[Mobihelp sharedInstance] leaveBreadcrumb:crumbDetails];
        [self callbackToJavascriptWithoutResultForCommand:command];
    }
    @catch(NSException* e) {
        [self callbackToJavascriptWithException:e ForCommand:command];
    }
}
-(void) setUserEmail :(CDVInvokedUrlCommand*)command {
    @try {
        NSString* emailAddress = [command argumentAtIndex:0];
        [[Mobihelp sharedInstance] setEmailAddress:emailAddress];
        [self callbackToJavascriptWithoutResultForCommand:command];
    }
    @catch (NSException* e) {
        [self callbackToJavascriptWithException:e ForCommand:command];
    }
    
}
-(void) setUserFullName :(CDVInvokedUrlCommand*)command {
    @try {
        NSString* username = [command argumentAtIndex:0];
        [[Mobihelp sharedInstance] setUserName:username];
        [self callbackToJavascriptWithoutResultForCommand:command];
    }
    @catch (NSException* e) {
        [self callbackToJavascriptWithException :e ForCommand:command];
    }
}
-(void) showAppRateDialog :(CDVInvokedUrlCommand*)command {
    [[Mobihelp sharedInstance] launchAppReviewRequest];
    [self callbackToJavascriptWithoutResultForCommand:command];
}
-(void) showConversations :(CDVInvokedUrlCommand*)command {
    [[Mobihelp sharedInstance] presentInbox:[self viewController]];
    [self callbackToJavascriptWithoutResultForCommand:command];
}
-(void) showFeedback :(CDVInvokedUrlCommand*)command {
    [[Mobihelp sharedInstance] presentFeedback:[self viewController]];
    [self callbackToJavascriptWithoutResultForCommand:command];
}
-(void) showSolutions :(CDVInvokedUrlCommand*)command {
    [[Mobihelp sharedInstance] presentSolutions:[self viewController]];
    [self callbackToJavascriptWithoutResultForCommand:command];
}
-(void) showSupport :(CDVInvokedUrlCommand*)command {
    [[Mobihelp sharedInstance] presentSupport:[self viewController]];
    [self callbackToJavascriptWithoutResultForCommand:command];
}




@end

File: src/android/MobihelpPlugin.java
//
//  MobihelpPlugin.java
//  FreshdeskSDK
//
//  Copyright (c) 2014 Freshdesk. All rights reserved.
//

package com.freshdesk.mobihelp.android;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaWebView;

import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONException;

import android.content.Context;
import android.util.Log;


import com.freshdesk.mobihelp.*;


public class MobihelpPlugin extends CordovaPlugin {

    private MobihelpConfig mobihelpConfig;
    private Context cordovaActivity;
    private boolean isInitialized = false;



    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView)
    {
        super.initialize(cordova, webView);
        cordovaActivity = cordova.getActivity();
    }


    @Override
    public boolean execute(String action, String rawargs,final CallbackContext callbackContext) throws JSONException
    {

        JSONArray args = new JSONArray(rawargs);

        //Utility functions
        if (action.equals("logToDevice")) {
            Log.v("MOBIHELP",args.getString(0));
            return true;
        }

        // Initialize function
        if (action.equals("init")) {

            JSONObject initArgs = new JSONObject(args.getString(0));

            String domain = initArgs.getString("domain"); domain = "http://".concat(domain);
            String appKey = initArgs.getString("appKey");
            String appSecret = initArgs.getString("appSecret");
            MobihelpConfig mobihelpConfig = new MobihelpConfig(domain, appKey, appSecret);

            //Set values (defaults are set in Javascript wrapper)
            mobihelpConfig.setAppStoreReviewUrl(initArgs.getString("androidAppStoreUrl"));
            mobihelpConfig.setAutoReplyEnabled(initArgs.getBoolean("autoReplyEnabled"));
            mobihelpConfig.setEnhancedPrivacyModeEnabled(initArgs.getBoolean("enhancedPrivacyModeEnabled"));
            mobihelpConfig.setFeedbackType(FeedbackType.valueOf(initArgs.getString("feedbackType")));
            mobihelpConfig.setLaunchCountForReviewPrompt(initArgs.getInt("launchCountForReviewPrompt"));
            mobihelpConfig.setPrefetchSolutions(initArgs.getBoolean("prefetchSolutions"));

            Mobihelp.init(cordovaActivity, mobihelpConfig);

            this.isInitialized = true;
            this.mobihelpConfig = mobihelpConfig;

            callbackContext.success();
            return true;
        }

        //MobihelpSDK must be initialozed before any other method can be called
        if(!isInitialized)
            throw new IllegalStateException("Mobihelp has not been initialized");



        //Mobihelp class functions
        if (action.equals("addCustomData")) {
            if(args.isNull(0))
                throw new IllegalArgumentException("Key for Custom Data not provided.");
            if(args.isNull(1))
                throw new IllegalArgumentException("Value for Custom Data not provided.");

            boolean sensitivity;
            if(args.isNull(2))
                sensitivity = false;
            else
                sensitivity = Boolean.parseBoolean(args.getString(2));

            Mobihelp.addCustomData(args.getString(0),args.getString(1),sensitivity);
            callbackContext.success();
            return true;
        }
        if (action.equals("clearBreadCrumbs")) {
            Mobihelp.clearBreadCrumbs(cordovaActivity);
            callbackContext.success();
            return true;
        }
        if (action.equals("clearCustomData")) {
            Mobihelp.clearCustomData(cordovaActivity);
            callbackContext.success();
            return true;
        }
        if (action.equals("clearUserData")) {
            Mobihelp.clearUserData(cordovaActivity);
            callbackContext.success();
            return true;
        }
        if (action.equals("getUnreadCount")) {
            int count = Mobihelp.getUnreadCount(cordovaActivity);
            callbackContext.success(count);
            return true;
        }
        if (action.equals("getUnreadCountAsync")) {
            UnreadUpdatesCallback unreadUpdatesCallback = new UnreadUpdatesCallback() {
                @Override
                public void onResult(MobihelpCallbackStatus mobihelpCallbackStatus, Integer integer)
                {
                    if (mobihelpCallbackStatus == MobihelpCallbackStatus.STATUS_SUCCESS)
                        callbackContext.success(integer);
                    else
                        callbackContext.error(mobihelpCallbackStatus.toString());
                }
            };
            Mobihelp.getUnreadCountAsync(cordovaActivity, unreadUpdatesCallback);
            return true;
        }
        if (action.equals("leaveBreadCrumb"))  {
            if(args.isNull(0))
                throw new IllegalArgumentException("CrumbDetails not provided.");
            Mobihelp.leaveBreadCrumb(args.getString(0));
            callbackContext.success();
            return true;
        }
        if (action.equals("setUserEmail")) {
            if(args.isNull(0))
                throw new IllegalArgumentException("Email not provided.");
            Mobihelp.setUserEmail(cordovaActivity, args.getString(0));
            callbackContext.success();
            return true;
        }
        if (action.equals("setUserFullName")) {
            if(args.isNull(0))
                throw new IllegalArgumentException("User Name not provided.");
            Mobihelp.setUserFullName(cordovaActivity, args.getString(0));
            callbackContext.success();
            return true;
        }
        if (action.equals("showAppRateDialog")) {
            Mobihelp.showAppRateDialog(cordovaActivity);
            callbackContext.success();
            return true;
        }
        if (action.equals("showConversations")) {
            Mobihelp.showConversations(cordovaActivity);
            callbackContext.success();
            return true;
        }
        if (action.equals("showFeedback")) {
            Mobihelp.showFeedback(cordovaActivity);
            callbackContext.success();
            return true;
        }
        if (action.equals("showSolutions")) {
            Mobihelp.showSolutions(cordovaActivity);
            callbackContext.success();
            return true;
        }
        if (action.equals("showSupport")) {
            Mobihelp.showSupport(cordovaActivity);
            callbackContext.success();
            return true;
        }

        return false;
    }


}